internal/auth/uaa_client.go

```diff
@@ -44,72 +44,110 @@ func NewUAAClient(
-			Token:   token,
-		}, nil
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := c.httpClient.Do(req)
+
+	if err != nil {
+		return fmt.Errorf("failed to get token key from UAA: %s", err)
 	}
 
-	form := url.Values{
-		"token": {trimBearer(token)},
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("got an invalid status code talking to UAA %v", resp.Status)
 	}
+	defer resp.Body.Close()
 
-	req, err := http.NewRequest("POST", c.uaa.String(), strings.NewReader(form.Encode()))
+	tokenKey, err := decodeTokenKey(resp.Body)
 	if err != nil {
-		log.Printf("failed to create UAA request: %s", err)
-		return Oauth2ClientContext{}, err
+		return err
 	}
-	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
-	req.SetBasicAuth(c.client, c.clientSecret)
 
-	start := time.Now()
-	resp, err := c.httpClient.Do(req)
-	c.storeLatency(float64(time.Since(start)))
+	if tokenKey.Value == "" {
+		return fmt.Errorf("received an empty token key from UAA")
+	}
+
+	block, _ := pem.Decode([]byte(tokenKey.Value))
+	if block == nil {
+		return fmt.Errorf("failed to parse PEM block containing the public key")
+	}
+
+	publicKeyInterface, err := x509.ParsePKIXPublicKey(block.Bytes)
 	if err != nil {
-		log.Printf("UAA request failed: %s", err)
-		return Oauth2ClientContext{}, err
+		return fmt.Errorf("error parsing public key: %s", err)
 	}
 
-	defer func() {
-		io.Copy(ioutil.Discard, resp.Body)
-		resp.Body.Close()
+	sharedKey, isRSAPublicKey := publicKeyInterface.(*rsa.PublicKey)
+	if !isRSAPublicKey {
+		return fmt.Errorf("did not get a valid RSA key from UAA: %s", err)
+	}
+
+	c.sharedKey = sharedKey
+
+	return nil
+}
+
+func (c *UAAClient) StartPeriodicTokenKeyRefresh(duration time.Duration) {
+	go func() {
+		for {
+			time.Sleep(duration)
+			err := c.GetTokenKey()
+			if err != nil {
+				c.log.Printf("background token refresh failed: %s", err)
+			}
+		}
 	}()
+}
 
-	if resp.StatusCode != http.StatusOK {
-		return Oauth2ClientContext{}, fmt.Errorf("Non-200 status code received from UAA")
+func (c *UAAClient) Read(token string) (Oauth2ClientContext, error) {
+	if token == "" {
+		return Oauth2ClientContext{}, errors.New("missing token")
+	}
+
+	if c.sharedKey == nil {
+		return Oauth2ClientContext{}, errors.New("missing shared key from UAA")
+	}
+
+	payload, _, err := jose.Decode(trimBearer(token), c.sharedKey)
+	if err != nil {
+		return Oauth2ClientContext{}, fmt.Errorf("failed to decode token: %s", err.Error())
 	}
 
-	uaaR, err = c.parseResponse(resp.Body)
+	decodedToken, err := decodeToken(strings.NewReader(payload))
 	if err != nil {
-		log.Printf("failed to parse UAA response body: %s", err)
-		return Oauth2ClientContext{}, err
+		return Oauth2ClientContext{}, fmt.Errorf("failed to unmarshal token: %s", err.Error())
 	}
 
-	c.tokenCache[token] = uaaR
+	if time.Now().After(decodedToken.ExpTime) {
+		return Oauth2ClientContext{}, fmt.Errorf("token is expired, exp = %s", decodedToken.ExpTime)
```

Comment:
Could we set the refresh time to something close to expire time?

Reply:
The refresh time is global, so it's independent of an individual token expiry. The guidance from UAA was to refresh "often".